content/docs/contributing/featuregates.md

```diff
@@ -0,0 +1,47 @@
+---
+Title: Implementing feature gates
+description: 'cert-manager contributing guide: Implementing feature gates'
+---
+
+As of v1 release cert-manager is considered stable. We aim to follow Kubernetes API compatibility policy when making API changes, see [API compatibility](../installation/api-compatibility.md) to avoid breaking users' existing cert-manager installations.
+This means that as developers we are somewhat limited in regards to changing existing behavior, i.e renaming or removing API elements or changing their behavior.
+
+New functionality that is not yet stable[^1] can still be added, but it needs to be placed behind a feature gate.
+
+## Feature gated API fields
+
+Feature gated API fields are implemented using `--feature-gates` flags of cert-manager [webhook](../cli/webhook.md) and [controller](../cli/controller.md).
+
+A feature gated API field is always visible to the user (i.e when running `kubectl explain <some-resource>`), but is only functional if the relevant feature is explicitly enabled via feature flags for both the webhook and controller.
+
+If a user attempts to apply a resource with the feature gated field set to a non-nil value, but the feature gate is not enabled, the resource will get rejected by the webhook validation.
+This mechanism differs from [the one that Kubernetes uses for feature gated API field implementation](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-field-in-existing-api-version) where the field will be simply set to nil if the feature gate is disabled. We chose to use webhook validation instead to make debugging easier for users who are attempting to use the feature gated field, but have forgotten to enable the feature gate.
```

Comment:
If an object is created using a feature gated thing while everything is enabled and then the controller/webhook are disabled, what happens when the controller restarts? Does anything flag for this?

Reply:
You would get errors in controller logs if there are resources with fields that are no longer valid. I've added a note to warn folks that resources need updating before disabling a previously enabled feature gate flags.